examples/classifier.c

```diff
@@ -15,6 +17,399 @@ float *get_regression_values(char **labels, int n)
+
+    ScoreMetric metric = evaluation_metrics[params.metric];
+
+
+    srand(time(0));
+    int seed = rand();
+
+    for(i = 0; i < ngpus; ++i){
+        srand(seed);
+#ifdef GPU
+        cuda_set_device(gpus[i]);
+#endif
+        nets[i] = load_network(cfgfile, weightfile, clear);
+        nets[i]->learning_rate *= ngpus;
+    }
+
+    srand(params.seed == -1 ? time(0) : params.seed);
+
+    network *net = nets[0];
+
+    int imgs = net->batch * net->subdivisions * ngpus;
+
+    list *options = read_data_cfg(datacfg);
+
+    char *backup_directory = option_find_str(options, "backup", "/backup/");
+    int tag = option_find_int_quiet(options, "tag", 0);
+    char *label_list = option_find_str(options, "labels", "data/labels.list");
+    char *train_list = option_find_str(options, "train", "data/train.list");
+    char *valid_list = option_find_str(options, "valid", "data/valid.list");
+    char *tree = option_find_str(options, "tree", 0);
+    if (tree) net->hierarchy = read_tree(tree);
+    int classes = option_find_int(options, "classes", 2);
+
+    // print network hiper-parameters
+    write_net_file(nets[0], backup_directory);
+
+    char **labels = 0;
+
+    if(!tag)
+        labels = get_labels(label_list);
+
+    list *plist = get_paths(train_list);
+    char **paths = (char **)list_to_array(plist);
+    int N = plist->size;
+
+    list *plist_valid = get_paths(valid_list);
+    int N_valid = plist_valid->size;
+
+
+    double time;
+
+    load_args args = {0};
+    args.w = net->w;
+    args.h = net->h;
+    args.threads = 32;
+    args.hierarchy = net->hierarchy;
+
+    args.min = net->min_ratio*net->w;
+    args.max = net->max_ratio*net->w;
+    args.angle = net->angle;
+    args.aspect = net->aspect;
+    args.exposure = net->exposure;
+    args.saturation = net->saturation;
+    args.hue = net->hue;
+    args.size = net->w;
+
+    args.paths = paths;
+    args.classes = classes;
+    args.n = imgs;
+    args.m = N;
+    args.labels = labels;
+    if (tag)
+        args.type = TAG_DATA;
+
+    else
+        args.type = CLASSIFICATION_DATA;
+
+    data train;
+    data buffer;
+    pthread_t load_thread;
+    args.d = &buffer;
+    load_thread = load_data(args);
+
+    int total_max_epochs = params.max_epochs / params.eval_epochs;
+    total_max_epochs = total_max_epochs > 1 ? total_max_epochs : 1;
+
+    float** y_score_train = new_mat(N, classes, sizeof(float));
+    float** y_score_valid = new_mat(N_valid, classes, sizeof(float));
+    int* y_true_train = malloc(N * sizeof(int));
+    int* y_true_valid = malloc(N_valid * sizeof(int));
+
+    float *train_accs = malloc(total_max_epochs * sizeof(float));
+    float *valid_accs = malloc(total_max_epochs * sizeof(float));
+    float best_acc = -1.0;
+    int bad_epochs = -1, update_epoch = 0;
+    int epoch = (*net->seen)/N, count = 0;
+    float epoch_loss = 0.0;
+
+
+    while(epoch < params.max_epochs && get_current_batch(net) < net->max_batches){
```

Comment:
There's a problem here: epoch may be < params.max_epochs at this point, but later on you update it (`epoch = *net->seen/N;`) and after it, use the epoch `cur_epoch` (which is calculated from it) to index an array. So you may end up writing to memory that you shouldn't.

So this must be restructured to account for that (perhaps break immediately if `epoch > max_epochs`?)